autofaiss/indices/distributed.py

```diff
@@ -261,8 +264,13 @@ def run(
     # broadcast the index bytes
     trained_index_bytes = get_bytes_from_index(faiss_index)
     broadcast_trained_index_bytes = ss.sparkContext.broadcast(trained_index_bytes)
-    nb_vectors = embedding_reader.count
-    nb_batches = math.ceil(nb_vectors / batch_size)  # use math.ceil to make sure that we cover every vector
+    sc = ss._jsc.sc()  # pylint: disable=protected-access
+    n_workers = len(sc.statusTracker().getExecutorInfos()) - 1
+
+    # maximum between the number of spark workers, 100M embeddings per task and the number of indices to keep
+    nb_batches = max(n_workers, int(embedding_reader.count / (10 ** 7)), nb_indices_to_keep)
```

Comment:
This is really good !
Now that we have less indices, maybe we can reduce the hard coded value 100 in:
`next_stage_folder = _merge_to_n_indices(spark_session=ss, n=100, src_folder=temporary_indices_folder,)`
like reduce it to 50 ? or something like `max(nb_batches // 2, nb_indices_to_keep)` ?
What do you think

Reply:
Yeah good question
I think let's keep it for another PR

But I'm thinking numbers that makes sense are something like
10000
100
1

So that each merge stage divides by 100 the number of indices

However indeed producing 10000 shards is unnecessary and costly
500
50
5
Can also make sense for example i think


Probably we should not do a 2 stage merge if the number of indices in the intermediary stage is very similar to the input stage

So I'm thinking having an intermediary stage count that is the number of worker could make sense. And if that number is not more than 2x the number of input indices then we don't do it

Reply:
ah thanks for asking about this piece of code

this is actually not working in some cases it seems, I'm hitting an issue where _batch_loader is producing some batches that contain no files at the end...
will fix

Reply:
Do you mean the new logic of less indices ?

Reply:
only this line

I fixed it like this https://github.com/criteo/autofaiss/pull/93/commits/1a99a2ddbef6d96d1f740deba687964ab600435d#diff-74d062bdcf37dcb982a51877295524a232ac150a84029cb600f0356561f6560dR273

Reply:
It looks like I can't see what was the previous version :(
What have you changed ?

Reply:
before:
```
    nb_batches = max(n_workers, int(embedding_reader.count / (10 ** 7)), nb_indices_to_keep)
    batch_size = math.ceil(embedding_reader.count / nb_batches)
```

now:
```
    estimated_nb_batches = max(n_workers, int(embedding_reader.count / (10 ** 7)), nb_indices_to_keep)
    batch_size = math.ceil(embedding_reader.count / estimated_nb_batches)
    nb_batches = math.ceil(embedding_reader.count / batch_size)
```

the problem with the previous version is that due to ceiling and the way the _batch_loader function, it was possible to have some empty indices batch

Reply:
ok, I see, thanks ! It looks good.